src/MagnumPlugins/ObjImporter/ObjImporter.cpp

```diff
@@ -115,387 +372,575 @@ ObjImporter::~ObjImporter() = default;
-        return std::nullopt;
+    if(objMat.specularTexture != -1) {
+        flags |= PhongMaterialData::Flag::SpecularTexture;
     }
-    if(!textureCoordinates.empty() && textureCoordinateIndices.size() != positionIndices.size()) {
-        CORRADE_INTERNAL_ASSERT(textureCoordinateIndices.size() < positionIndices.size());
-        Error() << "Trade::ObjImporter::mesh3D(): some texture coordinate indices are missing";
-        return std::nullopt;
+
+    PhongMaterialData* mat = new PhongMaterialData{
+            flags,
+            objMat.specularity};
+
+    if(objMat.ambientTexture == -1) {
+        mat->ambientColor() = objMat.ambient;
+    } else {
+        mat->ambientTexture() = objMat.ambientTexture;
     }
 
-    /* Merge index arrays, if there aren't just the positions */
-    std::vector<UnsignedInt> indices;
-    if(!normalIndices.empty() || !textureCoordinateIndices.empty()) {
-        std::vector<std::reference_wrapper<std::vector<UnsignedInt>>> arrays;
-        arrays.reserve(3);
-        arrays.emplace_back(positionIndices);
-        if(!normalIndices.empty()) arrays.emplace_back(normalIndices);
-        if(!textureCoordinateIndices.empty()) arrays.emplace_back(textureCoordinateIndices);
-        indices = MeshTools::combineIndexArrays(arrays);
-
-        /* Reindex data arrays */
-        try {
-            reindex(positionIndices, positions);
-            if(!normalIndices.empty()) reindex(normalIndices, normals.front());
-            if(!textureCoordinateIndices.empty()) reindex(textureCoordinateIndices, textureCoordinates.front());
-        } catch(...) {
-            /* Error message already printed */
-            return std::nullopt;
-        }
+    if(objMat.diffuseTexture == -1) {
+        mat->diffuseColor() = objMat.diffuse;
+    } else {
+        mat->diffuseTexture() = objMat.diffuseTexture;
+    }
 
-    /* Otherwise just use the original position index array. Don't forget to
-       check range */
+    if(objMat.specularTexture == -1) {
+        mat->specularColor() = objMat.specular;
     } else {
-        indices = std::move(positionIndices);
-        for(UnsignedInt i: indices) if(i >= positions.size()) {
-            Error() << "Trade::ObjImporter::mesh3D(): index out of range";
-            return std::nullopt;
-        }
+        mat->specularTexture() = objMat.specularTexture;
     }
 
-    return MeshData3D{*primitive, std::move(indices), {std::move(positions)}, std::move(normals), std::move(textureCoordinates), {}, nullptr};
+    return std::unique_ptr<AbstractMaterialData>(mat);
 }
 
+std::optional<ImageData2D> ObjImporter::doImage2D(UnsignedInt id) {
+    CORRADE_ASSERT(manager(), "Trade::ObjImporter::image2D(): the plugin must be instantiated with access to plugin manager in order to open image files", {});
+
+    std::unique_ptr<AbstractImporter> imageImporter = manager()->loadAndInstantiate("TgaImporter"); // probably AnyImageImporter would be the way to go here...
```

Comment:
Would it be possible to move Any*Importers into the magnum repository? Would that even make sense?